File: java/NatesHashTable.java
public class NatesHashTable {
	
	private NatesDataObject[] dataStorage;
	private int storageSize = 477;
	private String alphaMap = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ";
	
	public NatesHashTable() {
		dataStorage = new NatesDataObject[storageSize];
	}
	public NatesHashTable(double customSize) {
		storageSize = (int)customSize;
		
		dataStorage = new NatesDataObject[storageSize];
	}
	
	private int keyToIndex(String key) {
		int keyValue = 0;
		for (char c: key.toCharArray()) {
			keyValue += alphaMap.indexOf(c);
		}
		return keyValue%storageSize;
	}
	
	public boolean insert(NatesDataObject anObject) {
		String aKey = anObject.getKey();
		int index = keyToIndex(aKey);
		int start = index;
		
		while (dataStorage[index] != null) {
			index*=7;
			index %= storageSize;
			if (index == start) {
				System.out.println("No space left");
				return false;
			}
		}
		
		dataStorage[index] = anObject;
		return true;
	}
	
	public NatesDataObject find(String aKey) {
		int index = keyToIndex(aKey);
		int start = index;
		while (dataStorage[index].getKey() != aKey) {
			index*=7;
			index %= storageSize;
			if (index == start) return null;
		}
		return dataStorage[index];
	}
	
	public void printMap() {
		System.out.println("|-----------|");
		for (int i = 0; i < storageSize; i++) {
			if (dataStorage[i] == null) {
				System.out.println("| " + i + "   |");
			} else {
				System.out.println("| " + i + "   | " + dataStorage[i].getKey());
			}
		}
		System.out.println("|-----------|");
	}
	
}

File: javascript/NatesHashTable.js
(function() {
	
	'use strict';
	
	var alphaMap = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ";

	function NatesHashTable(customSize) {
		var _dataStorage = [],
			_storageSize = 477;
			
		if (customSize) {
			_storageSize = parseInt(customSize, 10);
		}
	
		function keyToIndex(key) {
			var keyValue = 0,
				c = 0,
				len = 0;
			for (c = 0, len = key.length; c < len; c++) {
				keyValue += alphaMap.indexOf(key[c]);
			}
			return keyValue % _storageSize;
		}
	
		function insert(anObject) {
			var aKey = anObject.getKey(),
				index = keyToIndex(aKey),
				start = index;
		
			// not exact because we want to match undefined too
			while (_dataStorage[index] != null) {
				index*=7;
				index %= _storageSize;
				if (index == start) {
					console.log("No space left");
					return false;
				}
			}
		
			_dataStorage[index] = anObject;
			return true;
		}
	
		function find(aKey) {
			var index = keyToIndex(aKey),
				start = index;
			
			while (_dataStorage[index].getKey() !== aKey) {
				index*=7;
				index %= _storageSize;
				if (index == start) return null;
			}
			return _dataStorage[index];
		}
		
		function toString() {
			var printStr = "|-----------|\n",
				i = 0;
		
			for (i = 0; i < _storageSize; i++) {
				// Inexact comparison, again because
				// we want to support undefined
				if (_dataStorage[i] == null) {
					printStr += "| " + i + "   |\n";
				} else {
					printStr += "| " + i + "   | " + _dataStorage[i].getKey() + "\n";
				}
			}
			printStr += "|-----------|\n";
			return printStr;
		}
		
		function printMap() {
			console.log(toString());
		}
		
		return {insert: insert, find: find, printMap: printMap, toString: toString};
	}

	window.NatesHashTable = NatesHashTable;

})();